client/src/main/java/io/getstream/chat/android/client/utils/observable/ChatEventsObservable.kt

```diff
@@ -0,0 +1,102 @@
+package io.getstream.chat.android.client.utils.observable
+
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.ConnectedEvent
+import io.getstream.chat.android.client.events.ConnectingEvent
+import io.getstream.chat.android.client.events.DisconnectedEvent
+import io.getstream.chat.android.client.events.ErrorEvent
+import io.getstream.chat.android.client.models.EventType
+import io.getstream.chat.android.client.socket.ChatSocket
+import io.getstream.chat.android.client.socket.ChatSocketService
+import io.getstream.chat.android.client.socket.SocketListener
+import java.util.Date
+
+internal class ChatEventsObservable(private val socket: ChatSocket) {
+
+    private val subscriptions = mutableSetOf<EventSubscription>()
+    private var eventsMapper = EventsMapper(this)
+
+    private fun onNext(event: ChatEvent) {
+        subscriptions.forEach { subscription ->
```

Comment:
I didn't want to create an intermediate list to iterate (which `filterNot` would do), on one hand for performance, but also because I wanted the check of whether one of these subscriptions is disposed to happen right before calling them (for example, to account for the case when one of them is disposed during iteration by the listener of a previous subscription).

Reply:
You can use subscriptions.sequence() if you don't want to create an intermediate list. 
🤔  Maybe you should use a synchronized block? I think you'll still have concurrence problem with this current approach if you to `onNext` 2x too fast.